File: py/wlearn/automl/_progressive.py
"""Progressive search matching JS automl/progressive.js."""

import math

import numpy as np

from ..errors import ValidationError
from ._common import detect_task, scorer_greater_is_better
from ._cv import stratified_k_fold, k_fold, get_scorer
from ._executor import Executor
from ._strategy_progressive import ProgressiveStrategy


class ProgressiveSearch:
    """Probe all candidates cheaply (1 fold + subsample),
    then promote top N to full K-fold evaluation."""

    def __init__(self, models, scoring=None, cv=5, seed=42, task=None,
                 n_iter=20, max_time_ms=0, promote_count=10,
                 probe_fraction=0.5):
        if not models:
            raise ValidationError(
                'ProgressiveSearch: at least one model is required')
        self._models = models
        self._scoring = scoring
        self._cv = cv
        self._seed = seed
        self._task = task
        self._n_iter = n_iter
        self._max_time_ms = max_time_ms
        self._promote_count = promote_count
        self._probe_fraction = probe_fraction
        self._leaderboard = None
        self._best_result = None

    def fit(self, X, y):
        task = self._task or detect_task(y)
        scoring = self._scoring or (
            'accuracy' if task == 'classification' else 'r2')
        greater_is_better = scorer_greater_is_better(scoring)

        # Probe folds: 2-fold, use only first
        if task == 'classification':
            probe_folds = stratified_k_fold(y, 2, do_shuffle=True,
                                            seed=self._seed)
        else:
            probe_folds = k_fold(len(y), 2, do_shuffle=True,
                                 seed=self._seed)
        single_fold = [probe_folds[0]]

        # Full folds for promoted candidates
        if task == 'classification':
            full_folds = stratified_k_fold(y, self._cv, do_shuffle=True,
                                           seed=self._seed + 1)
        else:
            full_folds = k_fold(len(y), self._cv, do_shuffle=True,
                                seed=self._seed + 1)

        strategy = ProgressiveStrategy(
            self._models,
            n_iter=self._n_iter,
            seed=self._seed,
            promote_count=self._promote_count,
            greater_is_better=greater_is_better,
            probe_fraction=self._probe_fraction,
        )

        # Phase 1: probe with 1-fold executor
        probe_time = (int(self._max_time_ms * 0.3)
                      if self._max_time_ms > 0 else 0)
        probe_executor = Executor(
            folds=single_fold,
            scoring=scoring,
            X=X, y=y,
            time_limit_ms=probe_time,
            seed=self._seed,
        )

        while strategy.phase == 'probe' and not strategy.is_done():
            if probe_executor.is_timed_out:
                break
            cand = strategy.next()
            if cand is None:
                break
            try:
                result = probe_executor.evaluate_candidate(
                    cand['candidateId'], cand['cls'], cand['params'],
                    cand.get('budget'))
                strategy.report(result)
            except Exception:
                strategy.report({
                    'candidateId': cand['candidateId'],
                    'meanScore': -float('inf'),
                    'foldScores': np.zeros(1),
                    'stdScore': 0,
                    'fitTimeMs': 0,
                    'nTrainUsed': 0,
                    'nTest': 0,
                })

        # Phase 2: full evaluation of promoted candidates
        full_time = (int(self._max_time_ms * 0.7)
                     if self._max_time_ms > 0 else 0)
        full_executor = Executor(
            folds=full_folds,
            scoring=scoring,
            X=X, y=y,
            time_limit_ms=full_time,
            seed=self._seed,
        )

        while not strategy.is_done():
            if full_executor.is_timed_out:
                break
            cand = strategy.next()
            if cand is None:
                break
            try:
                full_executor.evaluate_candidate(
                    cand['candidateId'], cand['cls'], cand['params'],
                    cand.get('budget'))
            except Exception:
                pass

        leaderboard = full_executor.leaderboard
        if leaderboard.length == 0:
            probe_lb = probe_executor.leaderboard
            if probe_lb.length == 0:
                raise ValidationError(
                    'ProgressiveSearch: no candidates were evaluated')
            self._leaderboard = probe_lb
        else:
            self._leaderboard = leaderboard

        self._best_result = self._leaderboard.best()
        return {'leaderboard': self._leaderboard,
                'bestResult': self._best_result}

    def refit_best(self, X, y):
        if self._best_result is None:
            raise ValidationError(
                'ProgressiveSearch: must call fit() first')
        best = self._best_result
        model_spec = None
        for m in self._models:
            if m['name'] == best['modelName']:
                model_spec = m
                break
        instance = model_spec['cls'].create(best['params'])
        instance.fit(X, y)
        return instance

    @property
    def leaderboard(self):
        return self._leaderboard

    @property
    def best_result(self):
        return self._best_result

File: packages/automl/src/progressive.js
import { stratifiedKFold, kFold, normalizeX, normalizeY,
  ValidationError } from '@wlearn/core'
import { Executor } from './executor.js'
import { ProgressiveStrategy } from './strategy-progressive.js'
import { detectTask, scorerGreaterIsBetter } from './common.js'

/**
 * Progressive search: probe all candidates cheaply (1 fold + subsample),
 * then promote top N to full K-fold evaluation.
 *
 * Faster than full random search when many candidates are weak.
 * The probe phase filters out bad configs quickly, saving time
 * for thorough evaluation of promising candidates.
 */
export class ProgressiveSearch {
  #models
  #opts
  #leaderboard = null
  #bestResult = null

  constructor(models, opts = {}) {
    if (!models || models.length === 0) {
      throw new ValidationError('ProgressiveSearch: at least one model is required')
    }
    this.#models = models
    this.#opts = {
      scoring: null,
      cv: 5,
      seed: 42,
      task: null,
      nIter: 20,
      maxTimeMs: 0,
      promoteCount: 10,
      probeFraction: 0.5,
      onProgress: null,
      ...opts,
    }
  }

  async fit(X, y) {
    const Xn = normalizeX(X)
    const yn = normalizeY(y)
    const task = this.#opts.task || detectTask(yn)
    const scoring = this.#opts.scoring || (task === 'classification' ? 'accuracy' : 'r2')
    const { cv, seed, nIter, maxTimeMs, promoteCount, probeFraction, onProgress } = this.#opts
    const greaterIsBetter = scorerGreaterIsBetter(scoring)

    // Probe folds: use only 1 fold for cheap screening
    const probeFolds = task === 'classification'
      ? stratifiedKFold(yn, 2, { shuffle: true, seed })
      : kFold(yn.length, 2, { shuffle: true, seed })
    // Use only the first fold for probing
    const singleFold = [probeFolds[0]]

    // Full folds for promoted candidates
    const fullFolds = task === 'classification'
      ? stratifiedKFold(yn, cv, { shuffle: true, seed: seed + 1 })
      : kFold(yn.length, cv, { shuffle: true, seed: seed + 1 })

    // Create strategy
    const strategy = new ProgressiveStrategy(this.#models, {
      nIter, seed, promoteCount, greaterIsBetter, probeFraction,
    })

    // Phase 1: probe with 1-fold executor
    const probeExecutor = new Executor({
      folds: singleFold,
      scoring,
      X: Xn,
      y: yn,
      timeLimitMs: maxTimeMs > 0 ? Math.floor(maxTimeMs * 0.3) : 0,
      seed,
      onProgress,
    })

    while (strategy.phase === 'probe' && !strategy.isDone()) {
      if (probeExecutor.isTimedOut) break
      const cand = strategy.next()
      if (cand === null) break
      try {
        const result = await probeExecutor.evaluateCandidate(cand)
        strategy.report(result)
      } catch {
        // Report a failing result so the strategy can count it
        strategy.report({
          candidateId: cand.candidateId,
          meanScore: -Infinity,
          foldScores: new Float64Array(1),
          stdScore: 0,
          fitTimeMs: 0,
          nTrainUsed: 0,
          nTest: 0,
        })
      }
    }

    // If probe timed out and strategy hasn't transitioned, force it
    if (strategy.phase === 'probe') {
      // Transition didn't happen (not all probes completed) - use what we have
      // The strategy's report method handles this
    }

    // Phase 2: full evaluation of promoted candidates
    const fullExecutor = new Executor({
      folds: fullFolds,
      scoring,
      X: Xn,
      y: yn,
      timeLimitMs: maxTimeMs > 0 ? Math.floor(maxTimeMs * 0.7) : 0,
      seed,
      onProgress,
    })

    while (!strategy.isDone()) {
      if (fullExecutor.isTimedOut) break
      const cand = strategy.next()
      if (cand === null) break
      try {
        await fullExecutor.evaluateCandidate(cand)
      } catch {
        // Skip failed candidates
      }
    }

    const leaderboard = fullExecutor.leaderboard
    if (leaderboard.length === 0) {
      // Fall back to probe results if no full evals completed
      const probeLeaderboard = probeExecutor.leaderboard
      if (probeLeaderboard.length === 0) {
        throw new ValidationError('ProgressiveSearch: no candidates were evaluated')
      }
      this.#leaderboard = probeLeaderboard
    } else {
      this.#leaderboard = leaderboard
    }

    this.#bestResult = this.#leaderboard.best()
    return { leaderboard: this.#leaderboard, bestResult: this.#bestResult }
  }

  async refitBest(X, y) {
    if (!this.#bestResult) {
      throw new ValidationError('ProgressiveSearch: must call fit() first')
    }
    const best = this.#bestResult
    const model = this.#models.find(m => m.name === best.modelName)
    const instance = await model.cls.create(best.params)
    const Xn = normalizeX(X)
    const yn = normalizeY(y)
    instance.fit(Xn, yn)
    return instance
  }

  get leaderboard() { return this.#leaderboard }
  get bestResult() { return this.#bestResult }
}
